api/logs/v1/labels_enforcer.go

```diff
@@ -67,7 +75,11 @@ func enforceValues(lm []*labels.Matcher, v url.Values) (values string, err error
 	expr.Walk(func(expr interface{}) {
 		switch le := expr.(type) {
 		case *logqlv2.StreamMatcherExpr:
```

Comment:
FWIW considering the observatorium/opa-openshift#15 draft PR it looks like the `opa.matcher` (considering we can only do `vector or vector` with stream label matcher in LogQL) is overloaded. The `label.Matchers` should maybe include a field that tells if they are going to be appended as label selectors in `StreamMatcherExpr` or as line filters., e.g.:
```go
	expr.Walk(func(expr interface{}) {
		switch le := expr.(type) {
		case *logqlv2.StreamMatcherExpr:
			le.AppendMatchers(lm)
		case *logqlv2.LineFilterExpr: // I believe our syntax does not support this case yet
			if mInfo.LogicalOp == logicalOr {
				le.AppendORMatchers(mInfo.Matchers)
			} else {
				le.AppendMatchers(mInfo.Matchers)
			}
		default:
			// Do nothing
		}
	})
```

Reply:
yes, or we do line filters anyway, regardless if it's vectors or not.

Reply:
Looking on the `StreamMatchersExpr.String()` implementation below I see that you circuit break with a closing curly brace in case of an or-Matcher and move forward with a line filter. This looks very to me. In the above sample I propose to clearly render or-Matchers as line filters, e.g. by appending a new line filter with or-s at the end of the query.

PS: I mention Line Filter above, but I am not sure if it the right expression. IIRC logical expressions on matchers are supported only on label filters and not line filters. Please consider the parser tests for the right expression.

Reply:
I tried your suggested approach at first but I hit an issue: if I remember correctly, Line filters "walker" is only invoked if there are actually line filters in the initial expression, it isn't invoked at all if there are none. Or maybe I didn't proceed like what you were thinking about... I can retry and let you see the  issue if you want

Reply:
I think the approach you suggest would work but requires a little more refactoring of the traversal pattern